### PR TITLE
Define mattr_* methods at caller's location

### PR DESCRIPTION
**Pull request by jhawthorn**
This makes the source_location of `mattr_accessor` be the location from which `mattr_accessor` was called instead of the source of `mattr_accessor` itself. This better matches `attr_accessor` and the other attributes we define (like `delegate`).

This also moves initialization of the class variable outside of the `class_eval`. Previously we were double-initializing it: once to set it to `nil` and a second time if it had a default value. I've combined these so that they're in the same place.

I was tempted to move these into a separate method (ie. `define_module_attribute_accessors`) which would both readers and writers and avoided duplication of setting the class variable, but it was a smaller change not to.

**Before:**

```
    [1] pry(main)> $ ActiveJob::Base.default_priority

    From: lib/active_support/core_ext/module/attribute_accessors.rb @ line 64:
    Owner: ActiveJob::QueuePriority::ClassMethods
    Visibility: public
    Number of lines: 3

    def #{sym}
      @@#{sym}
    end
```

**After:**

```
    [1] pry(main)> $ ActiveJob::Base.default_priority

    From: /home/jhawthorn/src/rails/activejob/lib/active_job/queue_priority.rb @ line 9:
    Owner: ActiveJob::QueuePriority::ClassMethods
    Visibility: public
    Number of lines: 1

    mattr_accessor :default_priority
```

### CODE REVIEW

**Comment by rafaelfranca** [comment]
It seems this broke Action View.

**Comment by jhawthorn** [comment]
This error was really confusing, order dependant, and took some debugging.

In `form_options_helper_test.rb` we were [defining the mattr_accessor](https://github.com/rails/rails/blob/master/actionview/test/template/form_options_helper_test.rb#L61) on a singleton instead of on the class itself.

The previous code did the wrong thing in two ways:
* `ActiveSupport::Timezone.fake_zone` was actually the _instance_ accessor, not the class level one. They have the same code, so this happened to work.
* `@@fake_zones` was being defined on `Module`, not `ActiveSupport::Timezone` 😱

Previously, because the eval did `@@fake_zones = nil` on the singleton that ended up being defined on `Module`. [Ruby's implementation explicitly skips singleton classes](https://github.com/ruby/ruby/blob/master/vm_insnhelper.c#L982-L1006) for all cvar instructions. This means that the accessor methods were also only looking on `Module`. The new code used `class_variable_set`, which, not being an instruction, dutifully sets the class variable directly on the singleton (which the doesn't work, since the accessors can only read from `Module`).

This was especially confusing in the way it broke our test suite because it would only fail when `form_options_helper_test.rb` was loaded, but other tests using `ActiveSupport::TimeZone` were run first, because once `ActiveSupport::TimeZone.fake_zones=` was assigned it would set that cvar on `Module`, which would "work".

I solved this by calling the `mattr_accessor :fake_zones` directly on the class. I'd prefer we didn't modify this class at all in that test, but that's a project for another time.

I've also made `mattr_accessor` error if it is called on a singleton class, so that hopefully nobody else has to figure all this out 😅.

TL;DR cvars are very confusing and we shouldn't be allowed to use them from singleton classes. `mattr_accessor` now errors if that's attempted.

**Review by rafaelfranca — APPROVED** [review]
